server/src/main/java/io/crate/execution/dml/upsert/TransportShardUpsertAction.java

```diff
@@ -319,12 +351,81 @@ static class IndexItemResponse {
     }
 
     @VisibleForTesting
-    protected IndexItemResponse insert(ShardUpsertRequest request,
+    protected IndexItemResponse insert(Indexer indexer,
+                                       ShardUpsertRequest request,
                                        ShardUpsertRequest.Item item,
                                        IndexShard indexShard,
                                        boolean isRetry,
                                        @Nullable ReturnValueGen returnGen,
                                        InsertSourceGen insertSourceGen) throws Exception {
+        if (insertSourceGen instanceof RawInsertSource
+                || insertSourceGen instanceof ValidatedRawInsertSource) {
+            return legacyInsert(request, item, indexShard, isRetry, returnGen, insertSourceGen);
+        }
+        final long version = request.duplicateKeyAction() == DuplicateKeyAction.OVERWRITE
+            ? Versions.MATCH_ANY
+            : Versions.MATCH_DELETED;
+
+        long startTime = System.nanoTime();
+        ParsedDocument parsedDoc = indexer.index(item);
+
+        // Remove this once replica uses indexer as well
+        item.source(parsedDoc.source());
+
+        if (!parsedDoc.newColumns().isEmpty()) {
+            var addColumnRequest = new AddColumnRequest(
+                RelationName.fromIndexName(indexShard.shardId().getIndexName()),
+                parsedDoc.newColumns(),
+                Map.of(),
+                new IntArrayList(0)
+            );
+            addColumnAction.execute(addColumnRequest).get();
```

Comment:
is it ok to call `get()` and make it sync here?

Reply:
It's not good but matches the previous behavior. Making the whole process async is a bigger topic.